orderer/common/multichannel/registrar.go

```diff
@@ -801,3 +801,60 @@ func (r *Registrar) loadJoinBlocks() map[string]*cb.Block {
 
 	return channelToBlockMap
 }
+
+func (r *Registrar) removeSystemChannel() error {
+	systemChannelID := r.systemChannelID
+	consensusType := r.systemChannel.ConsensusType()
+	if consensusType != "etcdraft" {
+		return errors.Errorf("cannot remove %s system channel: %s", consensusType, systemChannelID)
+	}
+
+	// halt the inactive chain registry
+	consenter := r.consenters["etcdraft"].(consensus.ClusterConsenter)
+	consenter.StopInactiveChainRegistry()
+
+	// halt all application channels
+	for channel, cs := range r.chains {
+		if channel == systemChannelID {
+			// system channel is halted after halting all app channels
+			delete(r.chains, channel)
+			continue
+		}
+		cs.Halt()
+
+		rl, err := r.ledgerFactory.GetOrCreate(channel)
+		if err != nil {
+			return errors.Wrapf(err, "could not retrieve ledger for channel: %s", channel)
+		}
+		configBlock := ConfigBlockOrPanic(rl)
+		isChannelMember, err := consenter.IsChannelMember(configBlock)
+		if err != nil {
+			return errors.Wrapf(err, "failed to determine channel membership for channel: %s", channel)
+		}
+		if !isChannelMember {
+			logger.Debugf("Not a member of channel %s, removing it", channel)
+			err := r.removeMember(channel, cs)
+			if err != nil {
+				return errors.Wrapf(err, "failed to remove channel: %s", channel)
+			}
+		}
+	}
+
+	// halt the system channel
+	r.systemChannel.Halt()
```

Comment:
Actually I would halt it first, not last. This minimizes the risk of a channel created after you stopped all the channels...  Also note that Halt is idempotent, this may simplify the code.

Reply:
Done. I tried simplifying things down but hit an issue re-halting the handwritten mock chain. I can definitely update the mock chain if we think the current state of the production code during application channel halting is too complicated.